src/producer/vector_producer.rs

```diff
@@ -1,57 +1,113 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
 use crate::producer::{ExtractCodeError, ResultsProducer};
 use crate::types::result::CheckResult;
 use reqwest::Client;
 use sentry_kafka_schemas::Schema;
 use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
 use tokio::task::JoinHandle;
+use tokio::time::{sleep, Duration};
+
+struct VectorRequestWorkerConfig {
+    vector_batch_size: usize,
+    endpoint: String,
+    retry_vector_errors_forever: bool,
+    max_retries: Option<u32>,
+    region: String,
+}
 
 pub struct VectorResultsProducer {
     schema: Schema,
     sender: UnboundedSender<Vec<u8>>,
+    pending_items: Arc<AtomicUsize>,
 }
 
 impl VectorResultsProducer {
     pub fn new(
         topic_name: &str,
         endpoint: String,
         vector_batch_size: usize,
+        retry_vector_errors_forever: bool,
+        max_retries: Option<u32>,
+        region: String,
     ) -> (Self, JoinHandle<()>) {
         let schema = sentry_kafka_schemas::get_schema(topic_name, None).unwrap();
         let client = Client::new();
+        let pending_items = Arc::new(AtomicUsize::new(0));
+
+        let config = VectorRequestWorkerConfig {
+            vector_batch_size,
+            endpoint,
+            retry_vector_errors_forever,
+            max_retries,
+            region,
+        };
 
         let (sender, receiver) = mpsc::unbounded_channel();
-        let worker = Self::spawn_worker(vector_batch_size, client, endpoint, receiver);
+        let worker = Self::spawn_worker(config, client, receiver, pending_items.clone());
 
-        (Self { schema, sender }, worker)
+        (
+            Self {
+                schema,
+                sender,
+                pending_items,
+            },
+            worker,
+        )
     }
 
     fn spawn_worker(
-        vector_batch_size: usize,
+        config: VectorRequestWorkerConfig,
         client: Client,
-        endpoint: String,
         mut receiver: UnboundedReceiver<Vec<u8>>,
+        pending_items: Arc<AtomicUsize>,
     ) -> JoinHandle<()> {
         tracing::info!("vector_worker.starting");
 
         tokio::spawn(async move {
-            let mut batch = Vec::with_capacity(vector_batch_size);
+            let mut batch = Vec::with_capacity(config.vector_batch_size);
 
             while let Some(json) = receiver.recv().await {
+                pending_items.fetch_sub(1, Ordering::SeqCst);
+                metrics::gauge!("producer.pending_items")
+                    .set(pending_items.load(Ordering::SeqCst) as f64);
```

Comment:
Nice, will be good to be able to keep an eye on this (and alert above some level)

Reply:
yup 100%